### PR TITLE
feat(HttpHandler): allow to skip auth on paths via callback

### PR DESCRIPTION
**Pull request by kostasrim**
* This PR allows the user of `HttpListenerBase` to register a `cb` that if it returns true, it will skip the `auth`.
* Added `username_` field to `HttpListenerBase` such that we can verify against a specific username

@romange I decided to slightly diverge from what we originally discussed, namely that the customization point will be the `auth` callback. For our use case, we don't really need that, as the only thing we want to do, is to skip the `auth` part on certain `paths`. Otherwise, we would need to also expose to the user, functions that deal with parsing of the `http` protocol and for our use case this is not really needed.

### CODE REVIEW

File: util/http/http_handler.cc
Comment (kostasrim) [inline util/http/http_handler.cc:409]: Whoops this will skip if the callback is not registered. Will fix it sec

File: util/http/http_handler.h
Comment (chakaz) [inline util/http/http_handler.h:96]: First, I didn't know you could specify parameter names in function signatures. Nice!
Second, I'd try to make this a unified API. Maybe 2 callbacks: `ShouldAuth(path)` and `TryAuth(user, pass)`
WDYT?
I guess that could be in another PR, but it'd be easier to pull helio once.

Comment (chakaz) [reply util/http/http_handler.h:96]: @royjacobson suggested to have a single API: `Auth(path, user, pass)`
It's a much better idea imho :)

Comment (kostasrim) [reply util/http/http_handler.h:96]: We don't really need `TryAuth(user, pass)` since I added `SetUsername` on the listener. That means, that both the `password` and `username` can be set by the consumers of this lib (in our case dragonfly). `SkipAuthForPath` allows to skip authorization on certain paths, e.g, `/metrics` (which is what we need) and `SetUsername` will allow us to auth against a pair of username/password set by the users of `HttpListenerBase`. 

Maybe I am missing something in your suggestion so I am all eyes :eyes: 

p.s. I had to google what WDYT is :rofl: It means what do you think right? :tong

**Comment by kostasrim** [comment]
The tests that fails is not related, and I see it's been failing for a while...

File: util/http/http_handler.h
Comment (royjacobson) [reply util/http/http_handler.h:96]: With a callback you can allow multiple users to authenticate to the same HTTP endpoint. I thought it's what we wanted for the ACL stuff, no?

Comment (kostasrim) [reply util/http/http_handler.h:96]: @royjacobson Good question and I am sorry I was not clear about this earlier on the daily. I had a discussion with @romange yesterday and we agreed that currently, admin should only be used by the `default` user. There is no current plan to integrate ACL with admin port or HTTP atm

Comment (chakaz) [reply util/http/http_handler.h:96]: It's still a better API for future expansion, no?

Comment (kostasrim) [reply util/http/http_handler.h:96]: I mean yes but I can't really foresee if this is something that we need and I would rather wait than introducing dead code (in a sense that is not used/executed anywhere)

Comment (romange) [reply util/http/http_handler.h:96]: I thought `Auth(path, user, pass)` is what we discussed - being able to whitelist a specific path and to support multiple "user/pass" pairs (where one of them is default/pass). If it somehow was not clear, I am sorry. 
so +1 to a `(path, user, pass)` functor

**Comment by romange** [comment]
Also, please rebase - the last commit should have fixed the CI

File: util/http/http_handler.h
Comment (kostasrim) [reply util/http/http_handler.h]: done

**Comment by kostasrim** [comment]
Do not merge this, I will reply here once I am sure this works well on dragonfly

File: util/http/http_handler.cc
Comment (romange) [inline util/http/http_handler.cc:385]: this logic won't work to provide public access to  `/metrics`, for example.

Comment (kostasrim) [reply util/http/http_handler.cc:385]: Why? I already tested this on DF and it works just fine? 

The callback would contain: 

```
[](auto path, auto username, auto password) {
  if(path == "/metrics") return true;
  //rest of auth here if the above branch is not true
}
```

https://github.com/dragonflydb/dragonfly/pull/1950/files#diff-1a804e4021a32c426f25ba8f4ad590c07f385dc90f01fe277fc9b8886ea90409R20

https://github.com/dragonflydb/dragonfly/pull/1950/files#diff-1a804e4021a32c426f25ba8f4ad590c07f385dc90f01fe277fc9b8886ea90409R20

Comment (kostasrim) [reply util/http/http_handler.cc:385]: Is there something I am missing?

Comment (romange) [reply util/http/http_handler.cc:385]: you won't reach the callback because the header will be missing for regular requests. `authenticated` will stay false.

Comment (romange) [reply util/http/http_handler.cc:385]: you should check how come it passed your tests

Comment (kostasrim) [reply util/http/http_handler.cc:385]: Nice catch! 
> you should check how come it passed your tests 
It passed the tests because I explicitly passed `username/password`. I added an extra test to cover this case. 

Again nice catch :pray:

Comment (chakaz) [reply util/http/http_handler.cc]: You could use the same API (i.e. with `path, user, pass`), and call it twice in the worst case. As long as the callback is fast, and it doesn't have any side effects, it might be an easier solution (or a simpler API)
But your call

File: util/http/http_handler.cc
Comment (chakaz) [inline util/http/http_handler.cc:411]: Any reason not to do:
```cpp
auto [path, query] = ParseQuery(target);
```
?

**Review by chakaz — APPROVED** [review]

File: util/http/http_handler.cc
Comment (romange) [reply util/http/http_handler.cc]: done

**Review by romange — APPROVED** [review]

**Review by chakaz — APPROVED** [review]

**Pull request merged**
